src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -643,55 +625,161 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
+                var arrayNotNullLabel = new object();
+
+                // T[]? array = PrivateImplementationDetails.cachingField;
+                // if (array is not null) goto arrayNotNull;
+                _builder.EmitOpCode(ILOpCode.Ldsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitBranch(ILOpCode.Brtrue, arrayNotNullLabel);
+
+                // array = new T[elementCount];
+                // RuntimeHelpers.InitializeArray(token, array);
+                // PrivateImplementationDetails.cachingField = array;
+                _builder.EmitOpCode(ILOpCode.Pop);
+                _builder.EmitIntConstant(elementCount);
+                _builder.EmitOpCode(ILOpCode.Newarr);
+                EmitSymbolToken(arrayType.ElementType, wrappedExpression.Syntax);
+                _builder.EmitArrayBlockInitializer(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+                _builder.EmitOpCode(ILOpCode.Dup);
+                _builder.EmitOpCode(ILOpCode.Stsfld);
+                _builder.EmitToken(cachingField, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
+
+                // arrayNotNullLabel:
+                // new ReadOnlySpan<T>(array)
+                _builder.MarkLabel(arrayNotNullLabel);
+                _builder.EmitOpCode(ILOpCode.Newobj, 0);
+                EmitSymbolToken(rosArrayCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
+            }
+
+            // Emit: new ReadOnlySpan<ElementType>(PrivateImplementationDetails.cachingField ??= new ElementType[] { ... constants ... })
+            bool tryEmitAsCachedArrayOfConstants(BoundArrayCreation arrayCreation, ArrayTypeSymbol arrayType, TypeSymbol elementType, NamedTypeSymbol spanType, bool used, BoundExpression? inPlaceTarget, out bool avoidInPlace)
+            {
+                avoidInPlace = false;
+                var initializer = arrayCreation.InitializerOpt;
+                if (initializer == null)
+                {
+                    return false;
+                }
+
+                var initializers = initializer.Initializers;
+                if (initializers.Any(static init => init.ConstantValueOpt == null))
+                {
+                    return false;
+                }
+
+                if (!tryGetReadOnlySpanArrayCtor(arrayCreation.Syntax, out var rosArrayCtor))
+                {
+                    return false;
+                }
+
+                Debug.Assert(!elementType.IsEnumType());
+
+                ImmutableArray<ConstantValue> constants = initializers.Select(static init => init.ConstantValueOpt!).ToImmutableArray();
+
+                if (constants.IsEmpty)
+                {
+                    emitEmptyReadonlySpan(spanType, arrayCreation, used, inPlaceTarget);
```

Comment:
>emitEmptyReadonlySpan [](http://example.com/codeflow?start=20&length=21)

Does it mean that empty spans are optimized regardless of element type? If so, consider doing this optimization earlier and in one place, rather than having two different places checking for the same condition.

Reply:
Moving that optimization earlier in the existing code affects some scenarios. Since that is not the purpose of the PR, I kept the original optimization in it's existing location.

Reply:
> Moving that optimization earlier in the existing code affects some scenarios. 

It doesn't look this way to me. We can discuss offline in more details